### ISSUE TITLE
Two phases preventing inference of WPs

### ISSUE THREAD

**Issue by mtzguido** [FStarLang/FStar#1499]
Here's a rough corner, assume we have this function
```f#
assume val unthunk (#a:Type) (#wpa : pure_wp a) ($f : (unit -> PURE a wpa)) : PURE a wpa
```
(the `$` qualifier on `f` is needed if we ever expect `wpa` to be inferred). Then, I would like to write
```f#
let _ = unthunk (fun () -> 2)
```
and have the WP inferred, but this fails with:
```
(Error 34) Computed type "Prims.PURE Prims.int (Prims.pure_return Prims.int 2)" does not match annotated type "Prims.PURE Prims.int (Prims.pure_null_wp Prims.int)", and no subtyping was allowed
```
since the implicit `wpa` was resolved, during the first phase, to `Prims.pure_null_wp Prims.int`. Which is correct, but too weak here. Disabling two phases helps this particular inference, since the more expressive `return` WP is inferred for the function, and the thing easily goes through.
```f#
#push-options "--use_two_phase_tc false"
let _ = unthunk (fun () -> 2)
#pop-options
```

Workaround: annotate WPs, but that's very very cumbersome.

**Comment by mtzguido** [comment]
Here's a crazy idea, what if phase 1 admitted queries instead of doing lax-checking? I tried it, and this example seems to work, as my other more complicated ones (still haven't run a significant test).

I guess doing so involves a performance hit, as we are computing WPs that are likely going to be dropped (but some of which we do want, like here). On the other hand, I'm often a bit confused about what lax-checking really is, and how it relates to admitting queries. This program shows a clear mismatch between both.

cc @nikswamy  @aseemr

**Comment by nikswamy** [comment]
In designing the 2-phase TC, our intention was to explicitly disentangle WP inference from inference of other types.

As such, the example that you have here is precisely the kind of thing that we decided would not be supported by the 2-phase TC since type inference explicitly relies here on WP inference.

We expected that code like this would not typecheck without annotations.

But, our experience so far in applying the 2-phase TC to our existing code suggests that this annotation requirement does not impact very much existing code. That said, we have also observed that the annotation requirement is burdensome in some code, notably in some code generated by Vale. In such code, we have explicitly disabled 2-phase TC there.

Making the first phase use `--admit_smt_queries true` may help in this example, but it will not be sufficient for other examples. The WP computed by the first phase is (intentionally) not always the same as the VC computed by the second phase. Any such mismatch would manifest in an error similar to the discrepancy you observed in your original post.
